internal/containerupdate/container_updater.go

```diff
@@ -2,13 +2,71 @@ package containerupdate
 
 import (
 	"context"
+	"fmt"
 	"io"
 
+	"github.com/windmilleng/tilt/internal/container"
+	"github.com/windmilleng/tilt/internal/docker"
+	"github.com/windmilleng/tilt/internal/k8s"
+	"github.com/windmilleng/tilt/internal/mode"
 	"github.com/windmilleng/tilt/internal/model"
 	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/internal/target"
 )
 
 type ContainerUpdater interface {
 	UpdateContainer(ctx context.Context, deployInfo store.DeployInfo,
 		archiveToCopy io.Reader, filesToDelete []string, cmds []model.Cmd, hotReload bool) error
+	SupportsSpecs(specs []model.TargetSpec) (supported bool, msg string)
+}
+
+// Differentiate between the CU we'll need for k8s updates (may be docker || synclet || exec)
+// and the one we'll need for DC updates (always docker)
+type K8sContainerUpdater ContainerUpdater
+type DCContainerUpdater ContainerUpdater
+
+func ProvideK8sContainerUpdater(kCli k8s.Client, dCli docker.Client, sm SyncletManager, env k8s.Env, updMode mode.UpdateMode, runtime container.Runtime) K8sContainerUpdater {
+	if updMode == mode.UpdateModeImage || updMode == mode.UpdateModeNaive {
+		return NewExplodingContainerUpdater()
+	}
+
+	if updMode == mode.UpdateModeKubectlExec {
+		return NewExecUpdater(kCli)
+	}
+
+	if updMode == mode.UpdateModeContainer || (runtime == container.RuntimeDocker && env.IsLocalCluster()) {
+		return NewDockerContainerUpdater(dCli, env, runtime)
+	}
+
+	if updMode == mode.UpdateModeSynclet || runtime == container.RuntimeDocker {
+		return NewSyncletUpdater(sm)
+	}
+
+	return NewExecUpdater(kCli)
+}
+
+func ProvideDCContainerUpdater(dCli docker.Client, updMode mode.UpdateMode, env k8s.Env, runtime container.Runtime) DCContainerUpdater {
+	if updMode == mode.UpdateModeImage || updMode == mode.UpdateModeNaive {
+		return NewExplodingContainerUpdater()
+	}
+
+	return NewDockerContainerUpdater(dCli, env, runtime)
+}
+
+// validation func shared by ExecUpdater and SyncletUpdater
+func specsAreOnlyImagesDeployedToK8s(specs []model.TargetSpec, updaterName string) (ok bool, msg string) {
```

Comment:
i don't think this is accurate? we should be able to update base images fine

Reply:
so it's based on [this check from the old syncletBaD](http://bit.ly/2Y1VtEb), and it makes sense -- how can you do a live update on a base image, if the essence of a live update is that it works on a running container?

anyway, figured out that integration tests were failing b/c I was passing the wrong set of specs in -- we care about the specs extracted for LiveUpdate, which is a subset of all the specs passed into the BaD

Reply:
ya, that's what i was worried about: i think the only way this API works is if SupportsSpecs() is called with an idiosyncratic and non-obvious set of inputs, and bugs are going to leak at that API boundary